### ISSUE TITLE
Make the appimage able to update itself

### ISSUE THREAD

**Issue by andy5995** [0ad-matters/0ad-appimage#9]
This would be nice. I don't know when I'll get to it though. Post on the ticket and let me know if you want to try tackling this issue. I *might* be able to help.

[Making AppImages updateable](https://docs.appimage.org/packaging-guide/optional/updates.html)

**Comment by andy5995** [comment]
Probably not worth it because there's so much data that gets changed between releases. Downloading an update would be about the same size as a release.

**Comment by rautamiekka** [comment]
Let's try it first, please. Surely even a very tight compression (like LZMA2 at Ultra) of the contents will still make the image correctly updateable.

- Updater: https://github.com/AppImageCommunity/AppImageUpdate
- Documentation for the updater: https://docs.appimage.org/packaging-guide/optional/updates.html

I can only help by trying to run the update as an end user, though.

**Comment by andy5995** [comment]
@rautamiekka Thanks for the the feedback. I'll reopen the issue.

**Comment by andy5995** [comment]
@ivan-hc said in #34 :

OFF-TOPIC: want to made it updateable?

If you use AppImagetool from https://github.com/AppImage/appimagetool, its enough you do
```
APPNAME="0ad"
REPO="0ad-appimage"
TAG="latest"
VERSION="0.27.0-2501310402"
UPINFO="gh-releases-zsync|$GITHUB_REPOSITORY_OWNER|$REPO|$TAG|*x86_64.AppImage.zsync"

ARCH=x86_64 ./appimagetool --comp zstd --mksquashfs-opt -Xcompression-level --mksquashfs-opt 20 \
	-u "$UPINFO" \
	./"$APP".AppDir "$APPNAME"_"$VERSION"-x86_64.AppImage
```

where GITHUB_REPOSITORY_OWNER is "0ad-matters" (if you run it in github actions workflows) and "$APP".AppDir is the AppDir, I don't know what's the name you use

**Comment by andy5995** [comment]
@ivan-hc Yes that might be much more convenient for some people if it were updateabe.

From your example, I'm not sure yet how it would work. I'll have to look more later and ask questions. If you want to open up a pull request to make the necessary changes, I could probably figure things out as we go, with your help.

**Comment by ivan-hc** [comment]
@andy5995 this is the same command I use in almost all my repositories (to be honest, @Samueru-sama pointed me on this)

All you need is:
1. the latest appimagetool from https://github.com/AppImage/appimagetool/releases/tag/continuous
2. your appdir

after you finished with the tool you use to create the AppImage, use appimagetool to convert the app directory (appdir) to an AppImage

the command in my comment does the follow:
- compresses the appdir using zstd with level 20
- includes th update info

that's all, a .zsync file with the same name of the AppImange will be created as well.

```
VERSION="0.27.0-2501310402"
ARCH=x86_64 ./appimagetool --comp zstd --mksquashfs-opt -Xcompression-level --mksquashfs-opt 20 \
	-u gh-releases-zsync|0ad-matters|0ad-appimage|latest|*x86_64.AppImage.zsync \
	./0ad.AppDir 0ad_"$VERSION"-x86_64.AppImage
```
I translated the line in the middle, the one referencing to the .zsync file.

If you want to allow forks, leave $GITHUB_REPOSITORY_OWNER instead of 0ad-matters. Its not difficult, this is the $UPINFO variable as before:

> ```
> APPNAME="0ad"
> REPO="0ad-appimage"
> TAG="latest"
> VERSION="0.27.0-2501310402"
> UPINFO="gh-releases-zsync|$GITHUB_REPOSITORY_OWNER|$REPO|$TAG|*x86_64.AppImage.zsync"
> 
> ARCH=x86_64 ./appimagetool --comp zstd --mksquashfs-opt -Xcompression-level --mksquashfs-opt 20 \
> 	-u "$UPINFO" \
> 	./"$APP".AppDir "$APPNAME"_"$VERSION"-x86_64.AppImage
> ```
> 
> where GITHUB_REPOSITORY_OWNER is "0ad-matters" (if you run it in github actions workflows) and "$APP".AppDir is the AppDir, I don't know what's the name you use

let me see the part of your script where you convert the AppDir to an AppImage.

**Comment by ivan-hc** [comment]
just noticed that you are using docker, maybe if the appimagetool can't be mounted, you should also use `--appimage-extract-and-run`... or alternativelly leave appimagetool itself taking care of conversion to AppImage directly, from Ubuntu.

**Comment by ivan-hc** [comment]
> If you want to open up a pull request to make the necessary changes, I could probably figure things out as we go, with your help.

@andy5995 I just forked the repository, please let me know how to only create the AppDir
```
linuxdeploy \
    -d $SOURCE_ROOT/build/resources/0ad.desktop \
    --icon-file=$SOURCE_ROOT/build/resources/0ad.png \
    --icon-filename=0ad \
    --executable $APPDIR/usr/bin/pyrogenesis \
    --library=/usr/lib/$ARCH-linux-gnu/libthai.so.0 \
    --custom-apprun=$WORKSPACE/AppRun \
    --appdir $APPDIR \
    --output appimage \
    --plugin gtk
fi
```
should I remove some flags from the above function? I suppose that `--output appimage` is the responsible of creating tha AppImage... isn't it?

**Comment by ivan-hc** [comment]
Just added a commit in my fork https://github.com/ivan-hc/0ad-appimage/commit/d23b7d80194ad22dda2d59a78b91f94db44f933b

Now I'll test if it works https://github.com/ivan-hc/0ad-appimage/actions/runs/13080646013

**Comment by ivan-hc** [comment]
If this has succes, wil also solve https://github.com/0ad-matters/0ad-appimage/issues/33

**Comment by ivan-hc** [comment]
I got an error also if the Appimage have been released.

Just added zsync and desktop-file-utils to have the .zsync file

The next workflow should do for our purpose https://github.com/ivan-hc/0ad-appimage/actions/runs/13081451548

**Comment by ivan-hc** [comment]
It seems that worked... but I made some mistake, the "latest" release also has the AppImages of my previous test

![Image](https://github.com/user-attachments/assets/eb4f0247-f766-467d-93a6-c1479bc078ec)

in this moment I'm downloading the AppImage, so I can test if update works.

**Comment by ivan-hc** [comment]
@andy5995 I don't know why, but with `appimageupdatetool` it does not works, on the contrary of `zsync` that works

https://github.com/user-attachments/assets/1e4c7e19-4cff-41ee-a29d-5284c07261f4

**Comment by ivan-hc** [comment]
Maybe the issue is that the target repository is a fork or something... I don't know why

@probonopd do you know why I get this error with `appimageupdatetool`?

```
Checking for updates...
terminate called after throwing an instance of 'appimage::update::updateinformation::UpdateInformationError'
  what():  GitHub API request failed: HTTP status 404, CURL error:
```

**Comment by ivan-hc** [comment]
> Maybe the issue is that the target repository is a fork or something... I don't know why

it seems so, from what I can read
- https://stackoverflow.com/questions/28936959/github-api-create-issues-return-404-not-found
- https://github.com/AppImageCommunity/AppImageUpdate/issues/184
- https://stackoverflow.com/questions/21337150/github-api-file-update-not-found

@andy5995 does your workflow using some kind of authentication?

**Comment by ivan-hc** [comment]
I think I found the issue

![Image](https://github.com/user-attachments/assets/73a67f5f-7b1f-445c-a0db-ef44c2c80e1d)

its time to made the PR

**Comment by ivan-hc** [comment]
@andy5995 I think this message should disappear once the AppImage and its zsync files are hosted on the right repo

![Image](https://github.com/user-attachments/assets/470e8ccc-598f-4430-94c2-b5379545bad9)

I start PR this

**Comment by andy5995** [comment]
> > If you want to open up a pull request to make the necessary changes, I could probably figure things out as we go, with your help.
> 
> [@andy5995](https://github.com/andy5995) I just forked the repository, please let me know how to only create the AppDir
> 
> ```
> linuxdeploy \
>     -d $SOURCE_ROOT/build/resources/0ad.desktop \
>     --icon-file=$SOURCE_ROOT/build/resources/0ad.png \
>     --icon-filename=0ad \
>     --executable $APPDIR/usr/bin/pyrogenesis \
>     --library=/usr/lib/$ARCH-linux-gnu/libthai.so.0 \
>     --custom-apprun=$WORKSPACE/AppRun \
>     --appdir $APPDIR \
>     --output appimage \
>     --plugin gtk
> fi
> ```
> 
> should I remove some flags from the above function? I suppose that `--output appimage` is the responsible of creating tha AppImage... isn't it?

As far as I know, those are all needed. And yes, the --output argument is required.

I didn't answer all your questions but I think you found answers and we are up to date. Is there anything I missed? I know you have a build running right now and very close to completing this.

**Comment by ivan-hc** [comment]
Yes, I have already completed some workflows... but I manually remove releases not to confuse myself when I must test them.

Anyway, next workflow should be successful https://github.com/ivan-hc/0ad-appimage/actions/runs/13083937340

About updates via `appimageupdatetool`, we only need to test this on the main branch... or at least, on a fork, with a different GITHUB_REPOSITORY_OWNER
https://github.com/ivan-hc/0ad-appimage/blob/b428372a8a92d4024134609fbccd3683548b7efd/make-appimage.sh#L222C8-L222C31

**Comment by andy5995** [comment]
Re-opening. We still need to a little testing and update docs.

**Comment by ivan-hc** [comment]
@andy5995 I was right, it needed to be hosted on the repo with the name of the owner

https://github.com/user-attachments/assets/d59d9c8c-09df-4b24-b472-6d7cd002fd20

...this means its working!

**Comment by ivan-hc** [comment]
@andy5995 same for the zsync file

https://github.com/user-attachments/assets/ba3ee967-4be0-4cc1-899d-ad8f6663fcb6

**Comment by ivan-hc** [comment]
Now, I suggest to keep only one per architecture, so people (me including) can download and update the right one.

NOTE, as I wrote in the script, TAG is "latest", so every time you release a new one in "latest" it will get updates from there. No pre-releases are targeted (for them I think you should add a dedicated script and workflow).

**Comment by ivan-hc** [comment]
> Now, I suggest to keep only one per architecture, so people (me including) can download and update the right one.

One also because my package manager "AM" targets this release with `wget` and `curl`

**Comment by ivan-hc** [comment]
@andy5995 since you just updated the AppImage, here you are the proof that updates are working as they should

https://github.com/user-attachments/assets/e6c5b150-8ae0-4e49-b547-dc2d8d3dca19

I think this issue is finally SOLVED!

**Comment by andy5995** [comment]
Yes, I experimented myself using the AppImageUpdate appimage https://github.com/AppImageCommunity/AppImageUpdate 

Should that be added to the readme or is there a different tool? I searched for appimageupdatetool with pacman and also in the AUR (I use Manjaro) but it's not available.

**Comment by ivan-hc** [comment]
> Should that be added to the readme or is there a different tool?

People can use both `appimageupdatetool` and `zsync`

> I searched for appimageupdatetool with pacman and also in the AUR (I use Manjaro) but it's not available.

Well, maybe its time for you to use something different ;) https://github.com/ivan-hc/AM

To install it system-wide
```
am -i appimageupdatetool
```
to install it locally in a rootless way
```
am -i --user appimageupdatetool
```
or
```
appman -i appimageupdatetool
```

**Comment by ivan-hc** [comment]
Since 0ad is a big AppImage, I keep it on a separated drive, so I can also integrate it insteado of installing it as another app listed... and it is updateable as well

https://github.com/user-attachments/assets/3357ada9-91db-4aa2-9836-223677b19622

...sorry for the long video, I wrote so many options and flags that I forgot that to update AppImages only integrated and NOT installed, you need to run
```
am -u --launcher
```

...and --launcher is also an option that has the same task of a normal AppImageLauncher or Gear Lever.... but in pure SHELL.

**Comment by andy5995** [comment]
> > Should that be added to the readme or is there a different tool?
> 
> People can use both `appimageupdatetool` and `zsync`
> 
> > I searched for appimageupdatetool with pacman and also in the AUR (I use Manjaro) but it's not available.
> 
> Well, maybe its time for you to use something different ;) https://github.com/ivan-hc/AM
> 

lol, ok! I'll try it out tomorrow. So late...

**Comment by ivan-hc** [comment]
> So late...

please, don't tell me... its 6 am here in Italy... big part of my work on github is done at  night (maybe it would be another reason because AM is called AM... its done at night) 😆

**Comment by andy5995** [comment]
> NOTE, as I wrote in the script, TAG is "latest", so every time you release a new one in "latest" it will get updates from there. No pre-releases are targeted (for them I think you should add a dedicated script and workflow).

@ivan-hc I understand there can be only one "latest" release, but I do I target pre-releases, since there can be more than one? Like when I want to target the latest prelease. Can I reference it by the tag or commit hash shown on the release page? Such as on here, where the tag and commit hash is shown?

**Comment by ivan-hc** [comment]
@andy5995 I think pre-release means anything, while "latest" is only one, regardless if the tag has a number.

I mean, you can also use $VERSION as $TAG, if it is different... or maybe you can use a dedicated tag "nightly" or "pre-release" if you want.

Remember, "latest" is only one and it can be replaced by a new "latest", on github.

On the contrary, other builds require a dedicated tag.

@Samueru-sama do you know how to do when the repo also conatains  several pre-releases?

**Comment by andy5995** [comment]
@ivan-hc I think you answered all my questions. It can basically just be whatever comes after '/tag/' in the release URL, e.g. https://github.com/0ad-matters/0ad-appimage/releases/tag/v0.0.26-alpha would be v0.0.26-alpha so easy enough.
